### PR TITLE
New deploy resource RPC

### LINKED ISSUE #8065: Create a new gRPC deploy request for DMN resources
**Description**

Currently, we have a gRPC request to deploy resources. But the naming is close to `processes` and doesn't include DMN resources. 

We should create a new request that is more agnostic to all kinds of resources (i.e. `DeployRequest` vs. `DeployProcessRequest`). 

* deploy request:
  * allow generic definitions of resources 
  * each resource has a name (i.e. its filename) and a content
* deploy response: 
  * contains the metadata of the deployed processes     
  * contains the metadata of the deployed decisions
* mark the previous request and response objects are deprecated  (note that we can't rename the objects because of the backward compatibility)

The current gRPC request and response look like this:

```
message DeployProcessRequest {
  // List of process resources to deploy
  repeated ProcessRequestObject processes = 1;
}

message ProcessRequestObject {
  // the resource basename, e.g. myProcess.bpmn
  string name = 1;
  // the process definition as a UTF8-encoded string
  bytes definition = 2;
}

message DeployProcessResponse {
  // the unique key identifying the deployment
  int64 key = 1;
  // a list of deployed processes
  repeated ProcessMetadata processes = 2;
}

message ProcessMetadata {
  // the bpmn process ID, as parsed during deployment; together with the version forms a
  // unique identifier for a specific process definition
  string bpmnProcessId = 1;
  // the assigned process version
  int32 version = 2;
  // the assigned key, which acts as a unique identifier for this process
  int64 processDefinitionKey = 3;
  // the resource name (see: ProcessRequestObject.name) from which this process was
  // parsed
  string resourceName = 4;
}
```

### LINKED ISSUE #8074: Handle the new gRPC deploy request
**Description**

We introduce a new gRPC deploy request for DMN resources. In the gateway, we should support the new request in the gRPC handler. The request and response should be mapped to the updated deployment record.

* handle the new gRPC deploy request in the gateway
  * the request message is similar to the previous message
  * fill the new properties of the response message with the updated deployment record
* keep handling the previous deploy request that is marked as deprecated   

blocked by #8065 
blocked by #8067

### PR DESCRIPTION
**Pull request by korthout**
## Description

Adds new gRPC DeployResource RPC, request and response messages to support non-process deployments, including decision and decision requirements for DMN resources. It is defined in a way that can be easily expanded.

This deprecates the existing DeployProcess RPC and messages.

It also expands the gateway with a new DeployResource RPC implementation.

## Related issues

closes #8065
closes #8074

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

### CODE REVIEW

**Comment by felix-mueller** [comment]
@korthout do we need to add anything to the Node JS Zeebe Client? maybe @jwulf can support updating it, but he would need to know what to change. maybe we can open an issue here https://github.com/camunda-community-hub/zeebe-client-node-js/issues if this is not already done by anyone else?
thank you

**Comment by Zelldon** [comment]
Similar to this https://github.com/camunda-community-hub/zeebe-client-csharp :)

**Comment by korthout** [comment]
@felix-mueller @jwulf @Zelldon Thanks for raising this. I was planning on announcing this on the forums to inform client maintainers after the changes are accepted. I can also open issues on the respective client repos we're aware of.

>do we need to add anything to the Node JS Zeebe Client

The Deploy Process RPC has been deprecated, but will still work as it did before because the public API maintains backward compatibility. When it comes to deploying processes and decisions, the client maintainers don't have to change anything at this time. 

If the clients want to support deploying DMN resources where the decision/drg metadata is available in the response, then they do need to implement the new RPC. Of course, we also want to encourage client maintainers to migrate to the new Deploy RPC, as we may drop the deprecated RPC in some future version (if we ever choose to break the backward compatibility for clients).

For now, I'd prefer to wait to inform client maintainers until these changes have been reviewed, so we make sure they don't have to make adjustments when we decide to change things during the review.

**Comment by korthout** [comment]
Hi @remcowesterhoud Welcome back 👋  Could you have a look at this PR?

**Review by remcowesterhoud — CHANGES_REQUESTED** [review]
Thanks @korthout! Your changes look good for the most part. What's mostly missing is some documentation.

On the topic of documentation. Should this deprecation also be documented in the cloud docs?

File: gateway-protocol/src/main/proto/gateway.proto [hunk 1]
```diff
@@ -396,8 +474,17 @@ service Gateway {
...
   rpc DeployProcess (DeployProcessRequest) returns (DeployProcessResponse) {
+    // deprecated, replaced by DeployResource rpc
+    option deprecated = true;
   }
 
+  /*
+    Deploys one or more resources (i.e. processes or decision models) to Zeebe.
+  */
+  rpc DeployResource (DeployRequest) returns (DeployResponse) {
+  }
```
Comment (korthout) [inline gateway-protocol/src/main/proto/gateway.proto:485]: I'm a bit unsure about the naming.

❓ @remcowesterhoud Should it be `DeployResource` or `Deploy`
❓ @remcowesterhoud Should it be `DeployResourceRequest` or `DeployRequest`
❓ @remcowesterhoud Should it be `DeployResourceResponse` or `DeployResponse`

Comment (remcowesterhoud) [reply gateway-protocol/src/main/proto/gateway.proto:485]: I personally like `DeployResource` over just `Deploy`. I don't have a very strong preference, but this gives a small amount of context about what you are deploying.

Comment (remcowesterhoud) [reply gateway-protocol/src/main/proto/gateway.proto:485]: And then also `DeployResourceRequest` and `DeployResourceResponse`. I think it's a good practice to keep those names aligned.

File: gateway-protocol/src/main/proto/gateway.proto [hunk 2]
```diff
@@ -396,8 +474,17 @@ service Gateway {
           - the process is invalid (e.g. an event-based gateway has an outgoing sequence flow to a task)
    */
   rpc DeployProcess (DeployProcessRequest) returns (DeployProcessResponse) {
+    // deprecated, replaced by DeployResource rpc
+    option deprecated = true;
   }
 
+  /*
+    Deploys one or more resources (i.e. processes or decision models) to Zeebe.
+  */
```
[the remark this answers is not in the document]
Comment (korthout) [reply gateway-protocol/src/main/proto/gateway.proto:483]: Thanks for raising this. I had a look at how this works currently.

If one resource is invalid the deployment is rejected. However, we have a bug: [valid resources in a rejected deployment still get deployed](https://github.com/camunda/zeebe/issues/5723). We should document the expected behavior, even if we have a bug that results in different behavior. We can then fix the bug to align with the documented behavior.

🔧 I'll make sure to document the expected behavior (atomic deployments).

File: gateway-protocol/src/main/proto/gateway.proto [hunk 3]
```diff
@@ -173,6 +207,50 @@ message ProcessMetadata {
...
+  int64 decisionRequirementsKey = 4;
+  // the dmn namespace, as parsed during deployment
+  // todo: debate this field (e.g. http://camunda.org/schema/1.0/dmn)
+  string namespace = 5;
+  // the resource name (see: Resource.name) from which this decision
+  // requirements was parsed
+  string resourceName = 6;
+  // whether this decision requirements was already deployed previously
+  // todo: debate this field
+  bool isDuplicate = 7;
```
[the remark this answers is not in the document]
Comment (korthout) [reply gateway-protocol/src/main/proto/gateway.proto:251]: I agree, I'll remove it for now

File: gateway-protocol/src/main/proto/gateway.proto [hunk 4]
```diff
@@ -173,6 +207,50 @@ message ProcessMetadata {
...
+  int64 decisionKey = 4;
+  // the dmn ID of the decision requirements graph that this decision is part
+  // of, as parsed during deployment
+  string dmnDecisionRequirementsId = 5;
+  // the assigned key of the decision requirements graph that this decision is
+  // part of
+  int64 decisionRequirementsKey = 6;
+  // whether this decision was already deployed previously
+  // todo: debate this field
+  bool isDuplicate = 7;
```
[the remark this answers is not in the document]
Comment (korthout) [reply gateway-protocol/src/main/proto/gateway.proto:229]: I agree, let's remove it for now

File: gateway-protocol/src/main/proto/gateway.proto (hunk 1 above)
Comment (korthout) [reply gateway-protocol/src/main/proto/gateway.proto:485]: Thanks, I like it 👍

**Comment by korthout** [comment]
@remcowesterhoud I've applied the hints. Please have another look.

>Should this deprecation also be documented in the cloud docs?

Yes, good point. I'll open an issue for it.

**Review by remcowesterhoud — APPROVED** [review]
Thanks for the changes! LGTM 🚀

**Comment by korthout** [comment]
bors merge

**Pull request merged**